subprojects/composite-builds/src/main/java/org/gradle/composite/internal/BuildControllers.java

```diff
@@ -37,6 +39,8 @@ interface BuildControllers extends Closeable {
      */
     BuildController getBuildController(BuildState build);
 
+    void withTasks(Consumer<? super Task> visitTask);
```

Comment:
These `withTasks()` methods were a hack to be able to tweak the test tasks after the work graph had been populated and/or loaded from the cache, and before execution starts.

I would probably move this so that it happens as part of the whole work graph calculation step, rather than something that happens afterwards to the "finalized" graph. In other words, the test task tweaks should happen as part of the work scoped by `BuildTreeWorkGraph.scheduleWork()`.  In particular, I think it makes sense to include this in the "build tree work graph finalization" step.

So, we might add something to `BuildTreeWorkGraph.Builder` to allow some "finalization actions" to be registered. These would run during the finalization step to do what ever and might take an `ExecutionPlan` as input.